### ISSUE TITLE
Custom loss function, different dimensions for y_pred and y_true

### ISSUE THREAD

**Issue by rodriguel** [keras-team/keras#10602]
Hello everyone! 
I'm trying to implement Yolo within Keras framework, and I'm having issue when it comes to the loss function. Indeed, when you look at this loss function ([see paper](https://pjreddie.com/media/files/papers/yolo.pdf), you can see that somehow, _y_true_ is going to depend on _y_pred_. In fact, in since each boxes in going to be affected as 'responsible' for a certain object at the first pass, we can only know what _y_true_ is after obtaining _y_pred_. 

So at first, what I wanted to do is the following :
- _y_pred_ is going to be the output of my ConvNet, of size (7,7, 35) (here, I'm only trying to detect and classify cats and dogs, using 5 anchor boxes per grid-cells, therefore the last dimension is indeed 5*(5 + 2) = 35. 
- Now, I wanted to define _y_true_ to be the list of objects in my image (I've created some class that contains the dimension of the box, its position and the classification (i.e. cats or dogs) of the object it contains).

Now the problem is, Keras seems to explicitly want _y_pred_ and _y_true_ to be of the same size, which is not really the case (there not even of the type...).

For information, here's an idea of my code...

     
`
 def lossFunction(y_true, y_pred):
	y_pred = np.array(y_pred)
	y_pred = y_pred.reshape((7,7,5,7))
	coord_c = 1
	noobj_c = 0.1

	loss = 0
	for bb in y_pred:

		# Fetching grid-cell responsible for object :
		x_center = bb.x
		y_center = bb.y 
		x_grid_responsible = int(x_center - x_center % (224 // 7)) / (224 // 7)
		y_grid_responsible = int(y_center - y_center % (224 // 7)) / (224 // 7)

		# Fetching anchor box in grid-cell responsible for object :
		boxes = list()
		boxess = y_pred[x_grid_responsible][y_grid_responsible][:][:]
		for i in range(5):
			# Getting all the anchors boxes of the grid-cell responsible for object bb
			box = AnchorBox(*boxess[i])
			boxes.append(box)

			# Getting the anchors box index k with highest-IOU to object bb
			responsible_anchor_box_idx = 0
			iou = 0
			for i in range(5):
				b_iou = box_iou(boxes[i], bb)
				if b_iou > iou:
					iou = b_iou
					responsible_anchor_box_idx = i

		if bb.class_name == 'cat':
			p0, p1 = 0, 1
		else:
			p0, p1 = 1, 0

		loss_center = coord_c * (K.square(y_pred[x_grid_responsible, y_grid_responsible, responsible_anchor_box_idx, 0] - bb.x) 
							+ K.square(y_pred[x_grid_responsible, y_grid_responsible, responsible_anchor_box_idx, 1] - bb.y))

		loss_dimension = coord_c * (K.square(K.sqrt(y_pred[x_grid_responsible, y_grid_responsible, responsible_anchor_box_idx, 2]) - K.sqrt(bb.w)) 
							+ K.square(K.sqrt(y_pred[x_grid_responsible, y_grid_responsible, responsible_anchor_box_idx, 3]) - K.sqrt(bb.h)))

		loss_confidence_o = K.square(y_pred[x_grid_responsible, y_grid_responsible, responsible_anchor_box_idx, 4] - iou)

		foo = 0
		for i in range(7):
			for j in range(7):
				for k in range(5):
					if not (i == x_grid_responsible and j == y_grid_responsible and k == responsible_anchor_box_idx):
						ab = AnchorBox(*y_pred[i,j,k])
						foo += K.square(y_pred[i,j,k,4] - box_iou(ab,bb))

		loss_confidence_no = noobj_c * foo

		loss_classification = (K.square(y_pred[x_grid_responsible, y_grid_responsible, responsible_anchor_box_idx,5] - p0)
							+ K.square(y_pred[x_grid_responsible, y_grid_responsible, responsible_anchor_box_idx, 6] - p1))

		return loss_center + loss_dimension + loss_confidence_o + loss_confidence_no + loss_classification `
